### ISSUE TITLE
`unreachable_from_main` should consider using the type as a usage

### ISSUE THREAD

**Issue by fzyzcjy** [dart-lang/linter#4471]
**Describe the issue**

I have some classes that are never instantiated. Instead, they are solely used in type signatures. For example,

```dart
void main() {
  final a = Point<GlobalCoordinate>();
  final a = Point<LocalCoordinate>();
}

class Point<Coordinate> {
  /* ... */
}

abstract class GlobalCoordinate {}

abstract class LocalCoordinate {}
```

![image](https://github.com/dart-lang/linter/assets/5236035/fb687a1c-19ac-4412-afde-ad8adcdf2385)

IMHO it should not warn, since it is indeed used.

**To Reproduce**
Ideally, please include an example.

**Expected behavior**

**Additional context**

**Comment by bwilkerson** [comment]
@srawlins

**Comment by srawlins** [comment]
I don't think they should count as used. The classes and the type arguments in that example can be safely removed.

**Comment by fzyzcjy** [comment]
@srawlins Sorry I cannot get it.

![image](https://github.com/dart-lang/linter/assets/5236035/b34bc9e7-c048-4934-9150-cfe9588a73b4)

**Comment by srawlins** [comment]
Thanks for the example. We can relax the restriction.

**Comment by fzyzcjy** [comment]
Thanks!
